File: lc/coreprep/implementation/practice_iterations/patterns_iterations_1/subsets_with_duplicates.cpp
#include <algorithm>
#include <iostream>
#include <list>
#include <queue>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;

// O(2n) / O(2n)
vector<vector<int>> subsets_with_duplicates(vector<int> &a) {
  sort(a.begin(), a.end());

  vector<vector<int>> res;
  res.push_back(vector<int>{});

  int end = -1;
  for (int i = 0; i < a.size(); i++) {
    vector<vector<int>> prev = res;
    int start = 0;

    if (i > 0 && a[i] == a[i - 1]) {
      start = end;
    }
    end = prev.size();

    for (int j = start; j < prev.size(); j++) {
      prev[j].push_back(a[i]);
      res.emplace_back(prev[j]);
    }
  }

  return res;
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n;
  cin >> tt;

  while (tt--) {
    cin >> n;

    vector<int> a(n);
    for (int i = 0; i < n; i++) {
      cin >> a[i];
    }

    auto res = subsets_with_duplicates(a);
    for (auto r : res) {
      for (int e : r) {
        cout << e << " ";
      }
      cout << "\n";
    }
  }

  return 0;
}

File: lc/coreprep/implementation/practice_iterations/patterns_iterations_1/SubsetsWithDuplicates.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class SubsetsWithDuplicates {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();

        int[] a = new int[n];
        for (int i = 0; i < n; i++) {
          a[i] = in.nextInt();
        }

        var res = subsetsWithDuplicates(a);
        for (var r : res) {
          for (int e : r) {
            System.out.printf("%d ", e);
          }
          System.out.println();
        }
      }
    }
  }

  // O(2n) / O(2n)
  public static List<List<Integer>> subsetsWithDuplicates(int[] a) {
    Arrays.sort(a);

    List<List<Integer>> res = new ArrayList<>();
    res.add(new ArrayList<>());

    int end = 0;
    for (int i = 0; i < a.length; i++) {
      int n = res.size();

      int start = (i > 0 && (a[i] == a[i - 1])) ? end : 0;
      end = n;

      for (int j = start; j < n; j++) {
        List<Integer> currSub = new ArrayList<>(res.get(j));
        currSub.add(a[i]);
        res.add(currSub);
      }
    }

    return res;
  }
}